java/org/apache/catalina/realm/MemoryRealm.java

```diff
@@ -167,23 +246,46 @@ public Principal authenticate(String username, String credentials) {
      * @param password User's password (clear text)
      * @param roles Comma-delimited set of roles associated with this user
      */
-    void addUser(String username, String password, String roles) {
+    void addUser(String username, String password, String roles, String fullname) {
 
         // Accumulate the list of roles for this user
-        List<String> list = new ArrayList<>();
+        Set<String> roleSet = new LinkedHashSet<>();
```

Comment:
Same as before, not related to this change.

Reply:
OK